### PR TITLE
Add Toggl extension

### PR DESCRIPTION
**Pull request by ThatNerdSquared**
## Description

Extension that allows you to work with Toggl timers. Currently implemented commands: start a recently used Toggl timer, stop a running Toggl timer.

## Type of change
- New extension

## Screencast

https://user-images.githubusercontent.com/72814106/149671989-dff7d615-e042-4391-a481-8eeae9028ecf.mov

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)

### CODE REVIEW

File: extensions/toggl-unofficial/package.json [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  "commands": [
+    {
+      "name": "startTimer",
+      "title": "Start Timer",
+      "description": "Start a Toggl timer from a list of unique timers used over the last 30 days.",
+      "mode": "view"
+    },
+    {
+      "name": "stopTimer",
+      "title": "Stop Running Timer",
```
Comment (mattisssa) [inline extensions/toggl-unofficial/package.json:18]:
```suggestion
      "title": "Stop Running Timer",
      "subtitle": "Toggl"
```

File: extensions/toggl-unofficial/package.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  "name": "toggl-unofficial",
+  "title": "Toggl (Unofficial)",
+  "description": "Start and stop Toggl timers! Not endorsed by Toggl in any way.",
+  "icon": "command-icon.png",
+  "author": "ThatNerd",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "startTimer",
+      "title": "Start Timer",
```
Comment (mattisssa) [inline extensions/toggl-unofficial/package.json:12]:
```suggestion
      "title": "Start Timer",
      "subtitle": "Toggl"
```

File: extensions/toggl-unofficial/package.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
+{
+  "$schema": "https://www.raycast.com/schemas/extension.json",
+  "name": "toggl-unofficial",
+  "title": "Toggl (Unofficial)",
+  "description": "Start and stop Toggl timers! Not endorsed by Toggl in any way.",
```
Comment (pernielsentikaer) [inline extensions/toggl-unofficial/package.json:5]:
```suggestion
  "name": "toggl",
  "title": "Toggl",
  "description": "Start and stop Toggl timers",
```

File: extensions/toggl-unofficial/README.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Toggl Unofficial
+
+This is a small extension I've built for [Raycast](https://raycast.com), designed to allow easy starting/stopping of Toggl timers. 
+
+I'll be submitting this extension to the Raycast Store shortly, so check back in a day or two for the published version.
```
Comment (pernielsentikaer) [inline extensions/toggl-unofficial/README.md:5]:
```suggestion
This is a small extension I've built for [Raycast](https://raycast.com), designed to allow easy starting/stopping of Toggl timers. 
```

File: extensions/toggl-unofficial/README.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+## Actions
+- Start Toggl Timer
+    - automatically retrieves unique time entries from the last 30 days
```
Comment (pernielsentikaer) [inline extensions/toggl-unofficial/README.md:9]:
```suggestion
- automatically retrieves unique time entries from the last 30 days
```

File: extensions/toggl-unofficial/README.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+- Stop Toggl Timer (not completed)
```
Comment (pernielsentikaer) [inline extensions/toggl-unofficial/README.md:10]: What about this, is this the one you added?

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @ThatNerdSquared 👋 

Thanks for your contribution 💯 

I left some feedback and suggestions for you – you should also rename the extensions folder to "toggl" instead.

Some more feedback:
- When a wrong token, it should dig gracefully (Showing a Toast and stop the loading spinner with an empty array as data)

aSwXF9gmIt6ZZ1rT

- Since I don't have any present timers, I'm a bit stuck here – Maybe you can get some ideas from #165 on how to allow start a new timer?

4OTIbfd3nzukPU9y

- If I use the “Stop Running timer” multiple times, it should also here die gracefully

FrsyPub7LdkR9crw

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by ThatNerdSquared** [comment]
Thanks for suggesting that I check out #165, @pernielsentikaer! I've been trying to find a clean way to let the user start a completely new timer and that gives me some useful ideas to play around with. Will report back after some experimentation...

**Comment by pernielsentikaer** [comment]
Awesome @ThatNerdSquared 💪

**Comment by franzwilhelm** [comment]
@ThatNerdSquared I've been working on a toggl track extension in parallell with you :O 

It supports creating time entries with tags, project and name
image

It also supports showing the current running time entry
image

And you can stop it
image

I suggest joining forces together to create the ultimate extension, so we don't have to publish two competing extensions to the store. WDYT @ThatNerdSquared :)

**Comment by pernielsentikaer** [comment]
Hi @franzwilhelm and @ThatNerdSquared 👋 

It sounds like a good plan to join forces with the Toggle extension, there are good idea in them both 💪

File: extensions/toggl-unofficial/README.md (hunk 6 above)
Comment (ThatNerdSquared) [reply extensions/toggl-unofficial/README.md]: Yup, fixed up some other things in the README too. Sorry about that.

**Comment by ThatNerdSquared** [comment]
@franzwilhelm also reached out to me on Slack, so I replied to him there. He has yet to respond back. 

For now, I've fixed up the items from your last review, @pernielsentikaer. Thanks for the feedback! Requesting another review now; if everything works out, I think we can ship this extension. I'll work with Franz in the future to add/adapt features so that this extension meshes with his workflow and ideas as well.

**Comment by pernielsentikaer** [comment]
Hi @ThatNerdSquared and @franzwilhelm 👋 

As you both are aware of, we can't have two identically extensions in the store, so we have to figure out which one should be the “base” of the extension, and then we join forces as already said. 

We do also have support for “contributors” in package.json, which is shown all the way around (also in the contributors store page)

Let me know your thoughts

**Comment by franzwilhelm** [comment]
Hi, I would be happy to make commits to @ThatNerdSquared, but as of now my extension has a lot more features. I also tested out this extension, and although I was able to create a new time entry, the following is missing from this extension that mine has to offer:

- **Support for all projects in all workspaces** (this one only lists projects from one of my workspaces). Hence, I was not able to create a time entry for the project I wanted.
- **Support for showing the current running time entry** with time elapsed, project and description. This is very useful to 1. know if you have any time entry to stop, and 2. to see the current status.
- **Support for adding tags** to new time entries

Also when I tried out the newest version of the extension I got this error. This is probably also related to the lack of multiple workspace support as it can't find the project of one of my recent time entries. That project is in a workspace it didn't fetch the projects for:
image

**Comment by ThatNerdSquared** [comment]
This is the part where I concede to Franz, I suppose. Feel free to close this PR @pernielsentikaer. Interesting that you can have competing PRs on this repo, I'll keep that in mind in the future if I see somebody else contributing an extension similar to mine 😉 

No need to add me as a contributor on the other extension, I will continue working on my own extension as a personal project locally, it makes a good learning project for TS/React.

**Comment by franzwilhelm** [comment]
Either way, I've been enjoying looking at your code after writing mine @ThatNerdSquared.  It's cool to see that many of the same issues have been solved in a similar manner. I would love if e.g. weekly/monthly stats were a part of the toggl track extension in the future. If you want to collaborate on that I would love to.

**Comment by pernielsentikaer** [comment]
Actually, I don't see it that way at all – this is mostly driven by the awesome community, and I want what's best for the community – I don't see it as a competition at any level.

It's not the first time it happened, we had 3 extensions doing the exact same within three hours, where everyone chipped in and made an awesome extension for the community.

I hope above makes sense, otherwise catch me in Slack for a conversation about it

**Comment by pernielsentikaer** [comment]
Closing in favour of #734

**Pull request closed without merging**